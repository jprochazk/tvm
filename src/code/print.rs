use std::fmt::Display;

use super::{CodeUnit, FnId, Function};
use crate::error::Location;
use crate::lex::Span;

pub struct DisplayModule<'a, 'src>(pub &'a CodeUnit, pub Option<&'src str>);
impl<'a, 'src> Display for DisplayModule<'a, 'src> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let DisplayModule(m, src) = self;

        writeln!(
            f,
            "{}",
            DisplayScriptFunction(None, &m.functions.main, *src)
        )?;
        for (id, fn_) in m.functions.script.iter().enumerate() {
            let id = Some(FnId(id as u16));
            writeln!(f, "{}", DisplayScriptFunction(id, fn_, *src))?;
        }
        Ok(())
    }
}

struct DisplayScriptFunction<'a, 'src>(Option<FnId>, &'a Function, Option<&'src str>);
impl<'a, 'src> Display for DisplayScriptFunction<'a, 'src> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let Self(id, fn_, src) = self;

        write!(f, "function {:?}", fn_.name)?;
        if let Some(id) = id {
            write!(f, " #{}", id.0)?;
        }
        writeln!(f, " {{")?;
        writeln!(f, "  registers: {} ({} params)", fn_.registers, fn_.params)?;
        write!(f, "  literals: ")?;
        if fn_.literals.is_empty() {
            writeln!(f, "[]")?;
        } else {
            writeln!(f, "[")?;
            for c in &fn_.literals {
                writeln!(f, "    {c:?}")?;
            }
            writeln!(f, "  ]")?;
        }
        writeln!(f, "  bytecode: ({} ops) [", fn_.bytecode.len())?;

        let mut max_len = 0;
        let mut instructions = Vec::with_capacity(fn_.bytecode.len());
        for op in fn_.bytecode.iter() {
            let s = op.to_string();
            max_len = std::cmp::max(max_len, s.len());
            instructions.push(s);
        }

        let mut prev_line_span = Span::empty();
        let mut instructions = instructions.into_iter().zip(fn_.spans.iter()).peekable();
        while let Some((instruction, span)) = instructions.next() {
            f.write_str("    ")?;
            'next: {
                if let Some(src) = src {
                    write!(f, "{instruction:max_len$}  // ")?;

                    if span.is_empty() {
                        break 'next;
                    }

                    let loc = Location::from_source_span(src, span);
                    if prev_line_span == loc.line_span() {
                        break 'next;
                    }
                    prev_line_span = loc.line_span();

                    f.write_str(src[loc.line_span()].trim())?;
                } else {
                    write!(f, "{instruction}")?;
                }
            }
            if instructions.peek().is_some() {
                f.write_str("\n")?;
            }
        }
        writeln!(f, "  ")?;
        writeln!(f, "  ]")?;
        writeln!(f, "}}")
    }
}
