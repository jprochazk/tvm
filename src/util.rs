use std::cell::RefCell;
use std::fmt::Display;

pub(crate) fn num_digits(v: usize) -> usize {
    use core::iter::successors;

    successors(Some(v), |&n| (n >= 10).then_some(n / 10)).count()
}

pub(crate) trait Discard {
    type Output;
    fn discard(self) -> Self::Output;
}

impl<T, E> Discard for Result<T, E> {
    type Output = Result<(), E>;

    fn discard(self) -> Self::Output {
        match self {
            Ok(_) => Ok(()),
            Err(e) => Err(e),
        }
    }
}

impl<T> Discard for Option<T> {
    type Output = Option<()>;

    fn discard(self) -> Self::Output {
        match self {
            Some(_) => Some(()),
            None => None,
        }
    }
}

pub struct Join<Iter, Sep>
where
    Iter: Iterator,
{
    iter: RefCell<Iter>,
    peek: RefCell<Option<Option<Iter::Item>>>,
    sep: Sep,
}

impl<Iter: Iterator, Sep> Join<Iter, Sep> {
    fn next(&self) -> Option<Iter::Item> {
        match self.peek.borrow_mut().take() {
            Some(item) => item,
            None => self.iter.borrow_mut().next(),
        }
    }

    fn has_next(&self) -> bool {
        let mut iter = self.iter.borrow_mut();
        let mut peek = self.peek.borrow_mut();
        peek.get_or_insert_with(|| iter.next());
        peek.is_some()
    }
}

impl<Iter, Sep> Display for Join<Iter, Sep>
where
    Iter: Iterator,
    <Iter as Iterator>::Item: Display,
    Sep: Display,
{
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        while let Some(item) = self.next() {
            write!(f, "{item}")?;
            if self.has_next() {
                write!(f, "{}", self.sep)?;
            }
        }
        Ok(())
    }
}

pub trait JoinIter: Sized + Iterator {
    fn join<Sep>(self, sep: Sep) -> Join<Self, Sep>;
}

impl<Iter> JoinIter for Iter
where
    Iter: Sized + Iterator,
{
    fn join<Sep>(self, sep: Sep) -> Join<Self, Sep> {
        Join {
            iter: RefCell::new(self),
            peek: RefCell::new(None),
            sep,
        }
    }
}
