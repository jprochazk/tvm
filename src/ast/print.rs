use std::fmt::{Debug, Display, Formatter, Result};

impl Debug for super::Ast<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("Ast")
            .field("decls", &self.decls)
            .field("top_level", &self.top_level)
            .finish()
    }
}

impl Debug for super::decl::Decl<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        Debug::fmt(&self.kind, f)
    }
}

impl Debug for super::decl::DeclKind<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        match self {
            Self::Fn(arg0) => Debug::fmt(arg0, f),
            Self::Type(arg0) => Debug::fmt(arg0, f),
        }
    }
}

impl Debug for super::decl::Fn<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("Fn")
            .field("name", &self.name)
            .field("params", &self.params)
            .field("ret", &self.ret)
            .field("body", &self.body)
            .finish()
    }
}

impl Debug for super::decl::Body<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        match self {
            Self::Extern => write!(f, "Extern"),
            Self::Block(arg0) => f.debug_tuple("Block").field(arg0).finish(),
        }
    }
}

impl Debug for super::decl::TypeDef<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("Type")
            .field("name", &self.name)
            .field("fields", &self.fields)
            .finish()
    }
}

impl Debug for super::decl::Fields<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        match self {
            Self::Extern => write!(f, "Extern"),
            Self::Named(arg0) => f.debug_tuple("Named").field(arg0).finish(),
        }
    }
}

impl Debug for super::decl::Field<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("Field")
            .field("name", &self.name)
            .field("ty", &self.ty)
            .finish()
    }
}

impl Debug for super::stmt::Stmt<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        Debug::fmt(&self.kind, f)
    }
}

impl Debug for super::stmt::StmtKind<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        match self {
            Self::Let(arg0) => Debug::fmt(arg0, f),
            Self::Loop(arg0) => Debug::fmt(arg0, f),
            Self::Expr(arg0) => Debug::fmt(arg0, f),
        }
    }
}

impl Debug for super::stmt::Let<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("Let")
            .field("name", &self.name)
            .field("ty", &self.ty)
            .field("init", &self.init)
            .finish()
    }
}

impl Debug for super::stmt::Loop<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("Loop").field("body", &self.body).finish()
    }
}

impl Display for super::ty::Ty<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        use super::ty::TyKind as Ty;
        match &self.kind {
            Ty::Empty => f.write_str("_"),
            Ty::Named(ty) => f.write_str(ty.name.lexeme),
        }
    }
}

impl Debug for super::ty::Ty<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        Debug::fmt(&self.kind, f)
    }
}

impl Debug for super::ty::TyKind<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        match self {
            Self::Empty => write!(f, "Empty"),
            Self::Named(arg0) => f.debug_tuple("Named").field(arg0).finish(),
        }
    }
}

impl Debug for super::ty::Named<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("Named").field("name", &self.name).finish()
    }
}

impl Debug for super::expr::Expr<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        Debug::fmt(&self.kind, f)
    }
}

impl Debug for super::expr::ExprKind<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        match self {
            Self::Return(arg0) => Debug::fmt(arg0, f),
            Self::Break => write!(f, "Break"),
            Self::Continue => write!(f, "Continue"),
            Self::Block(arg0) => Debug::fmt(arg0, f),
            Self::If(arg0) => Debug::fmt(arg0, f),
            Self::Binary(arg0) => Debug::fmt(arg0, f),
            Self::Unary(arg0) => Debug::fmt(arg0, f),
            Self::Primitive(arg0) => Debug::fmt(arg0, f),
            Self::Array(arg0) => Debug::fmt(arg0, f),
            Self::UseVar(arg0) => Debug::fmt(arg0, f),
            Self::UseField(arg0) => Debug::fmt(arg0, f),
            Self::UseIndex(arg0) => Debug::fmt(arg0, f),
            Self::AssignVar(arg0) => Debug::fmt(arg0, f),
            Self::AssignField(arg0) => Debug::fmt(arg0, f),
            Self::AssignIndex(arg0) => Debug::fmt(arg0, f),
            Self::Call(arg0) => Debug::fmt(arg0, f),
            Self::MethodCall(arg0) => Debug::fmt(arg0, f),
        }
    }
}

impl Debug for super::expr::Return<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("Return")
            .field("value", &self.value)
            .finish()
    }
}

impl Debug for super::expr::If<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("If")
            .field("branches", &self.branches)
            .field("tail", &self.tail)
            .finish()
    }
}

impl Debug for super::expr::Binary<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("Binary")
            .field("left", &self.lhs)
            .field("op", &self.op)
            .field("right", &self.rhs)
            .finish()
    }
}

impl Debug for super::expr::Unary<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("Unary")
            .field("op", &self.op)
            .field("right", &self.rhs)
            .finish()
    }
}

impl Debug for super::expr::Primitive<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        match self {
            Self::Int(arg0) => f.debug_tuple("Int").field(arg0).finish(),
            Self::Num(arg0) => f.debug_tuple("Num").field(arg0).finish(),
            Self::Bool(arg0) => f.debug_tuple("Bool").field(arg0).finish(),
            Self::Str(arg0) => f.debug_tuple("Str").field(arg0).finish(),
        }
    }
}

impl Debug for super::expr::Array<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("Array").field("items", &self.items).finish()
    }
}

impl Debug for super::expr::UseVar<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("UseVar").field("name", &self.name).finish()
    }
}

impl Debug for super::expr::UseField<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("UseField")
            .field("parent", &self.parent)
            .field("name", &self.name)
            .finish()
    }
}

impl Debug for super::expr::UseIndex<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("UseIndex")
            .field("parent", &self.parent)
            .field("key", &self.key)
            .finish()
    }
}

impl Debug for super::expr::AssignVar<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("AssignVar")
            .field("name", &self.name)
            .field("op", &self.op)
            .field("value", &self.value)
            .finish()
    }
}

impl Debug for super::expr::AssignField<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("AssignField")
            .field("parent", &self.parent)
            .field("name", &self.name)
            .field("op", &self.op)
            .field("value", &self.value)
            .finish()
    }
}

impl Debug for super::expr::AssignIndex<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("AssignIndex")
            .field("parent", &self.parent)
            .field("key", &self.key)
            .field("op", &self.op)
            .field("value", &self.value)
            .finish()
    }
}

impl Debug for super::expr::Call<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("Call")
            .field("callee", &self.callee)
            .field("args", &self.args)
            .finish()
    }
}

impl Debug for super::expr::MethodCall<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("MethodCall")
            .field("receiver", &self.receiver)
            .field("method", &self.method)
            .field("args", &self.args)
            .finish()
    }
}

impl Display for super::BinaryOp {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        use super::BinaryOp as Op;
        f.write_str(match self {
            Op::Add => "+",
            Op::Sub => "-",
            Op::Mul => "*",
            Op::Div => "/",
            Op::Rem => "%",
            Op::Pow => "**",
            Op::Eq => "==",
            Op::Ne => "!=",
            Op::Gt => ">",
            Op::Lt => "<",
            Op::Ge => ">=",
            Op::Le => "<=",
            Op::And => "&&",
            Op::Or => "||",
            Op::Opt => "??",
        })
    }
}

impl Debug for super::BinaryOp {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        write!(f, "Op({self})")
    }
}

impl Display for super::UnaryOp {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        use super::UnaryOp as Op;
        f.write_str(match self {
            Op::Minus => "-",
            Op::Not => "!",
            Op::Opt => "?",
        })
    }
}

impl Debug for super::UnaryOp {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        write!(f, "Op({self})")
    }
}

impl Debug for super::Param<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_tuple("Param")
            .field(&self.name)
            .field(&self.ty)
            .finish()
    }
}

impl Debug for super::Branch<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_tuple("Branch")
            .field(&self.cond)
            .field(&self.body)
            .finish()
    }
}

impl Debug for super::Arg<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut t = f.debug_tuple("Arg");
        match &self.key {
            Some(key) => t.field(key).field(&self.value).finish(),
            None => t.field(&self.value).finish(),
        }
    }
}

impl Debug for super::Block<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Block")
            .field("body", &self.body)
            .field("tail", &self.tail)
            .finish()
    }
}

impl Display for super::Ident<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.write_str(self.lexeme)
    }
}

impl Debug for super::Ident<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        write!(f, "Ident({:?})", self.lexeme)
    }
}
