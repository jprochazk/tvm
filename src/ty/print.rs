use std::fmt::{Display, Write};

use super::{Hir, Type};
use crate::ast::*;

#[derive(Clone, Copy)]
struct Indent {
    level: usize,
}

impl Indent {
    fn next(mut self) -> Self {
        self.level += 2;
        self
    }
}

impl Display for Indent {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let indent = self.level;
        write!(f, "{:indent$}", "")
    }
}

impl Display for Hir<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let i = Indent { level: 0 };

        /* for decl in &self.decls {
          match &decl.kind {
            decl::DeclKind::Fn(decl) => {
              print_function_header(f, i, decl)?;
              print_block(f, i, &decl.body)?;
            }
          }
        } */

        {
            let b = &self.top_level;
            for stmt in &b.body {
                write!(f, "{i}")?;
                print_stmt(f, i, stmt)?;
                writeln!(f, ";")?;
            }
            if let Some(tail) = &b.tail {
                print_expr(f, i, tail)?;
            }
        }

        Ok(())
    }
}

fn print_function_header(
    f: &mut impl Write,
    i: Indent,
    decl: &decl::Fn<'_, Type>,
) -> std::fmt::Result {
    writeln!(f, "fn {}(", decl.name)?;
    for param in &decl.params {
        let i = i.next();
        writeln!(f, "{i}{}: {},", param.name, param.ty)?;
    }
    write!(f, "{i}) ")?;
    if let Some(ret) = &decl.ret {
        write!(f, "-> {ret} ")?;
    }
    print_block(f, i, &decl.body)?;
    Ok(())
}

fn print_block(f: &mut impl Write, i: Indent, b: &Block<'_, Type>) -> std::fmt::Result {
    writeln!(f, "{{")?;
    {
        let i = i.next();
        for stmt in &b.body {
            write!(f, "{i}")?;
            print_stmt(f, i, stmt)?;
            writeln!(f, ";")?;
        }
        if let Some(tail) = &b.tail {
            print_expr(f, i, tail)?;
        }
    }
    writeln!(f, "}}")?;

    Ok(())
}

fn print_stmt(f: &mut impl Write, i: Indent, s: &Stmt<'_, Type>) -> std::fmt::Result {
    match &s.kind {
        stmt::StmtKind::Let(s) => {
            write!(f, "let {}", s.name)?;
            if let Some(ty) = &s.ty {
                write!(f, ": {}", ty)?;
            }
            write!(f, " = ")?;
            print_expr(f, i, &s.init)?;
        }
        stmt::StmtKind::Loop(s) => {
            write!(f, "loop ")?;
            print_block(f, i, &s.body)?;
        }
        stmt::StmtKind::Expr(s) => {
            print_expr(f, i, s)?;
        }
    }

    Ok(())
}

fn print_expr(f: &mut impl Write, i: Indent, e: &Expr<'_, Type>) -> std::fmt::Result {
    match &e.kind {
        expr::ExprKind::Return(e) => {
            write!(f, "return ")?;
            if let Some(e) = &e.value {
                print_expr(f, i, e)?;
            }
        }
        /* expr::ExprKind::Yield(e) => {
          write!(f, "yield ")?;
          if let Some(e) = &e.value {
            print_expr(f, i, e)?;
          }
        } */
        expr::ExprKind::Break => {
            write!(f, "break")?;
        }
        expr::ExprKind::Continue => {
            write!(f, "continue")?;
        }
        expr::ExprKind::Block(e) => {
            print_block(f, i, e)?;
        }
        expr::ExprKind::If(e) => {
            let mut first = true;
            for branch in &e.branches {
                if !first {
                    write!(f, " else ")?;
                }
                first = false;
                write!(f, "if ")?;
                print_expr(f, i, &branch.cond)?;
                write!(f, " ")?;
                print_block(f, i, &branch.body)?;
            }
            if let Some(tail) = &e.tail {
                write!(f, " else ")?;
                print_block(f, i, tail)?;
            }
        }
        expr::ExprKind::Binary(e) => {
            print_expr(f, i, &e.lhs)?;
            write!(f, " {} ", e.op)?;
            print_expr(f, i, &e.rhs)?;
        }
        expr::ExprKind::Unary(e) => {
            write!(f, "{} ", e.op)?;
            print_expr(f, i, &e.rhs)?;
        }
        expr::ExprKind::Primitive(e) => match &**e {
            expr::Primitive::Int(v) => write!(f, "{v}")?,
            expr::Primitive::Num(v) => write!(f, "{v}")?,
            expr::Primitive::Bool(v) => write!(f, "{v}")?,
            expr::Primitive::Str(v) => write!(f, "{v:?}")?,
        },
        expr::ExprKind::Array(e) => match &**e {
            expr::Array::Csv(a) => {
                write!(f, "[")?;
                for e in a {
                    let i = i.next();
                    print_expr(f, i, e)?;
                    write!(f, ",")?;
                }
                write!(f, "]")?;
            }
            expr::Array::Len(e, l) => {
                write!(f, "[")?;
                print_expr(f, i, e)?;
                write!(f, ";")?;
                print_expr(f, i, l)?;
                write!(f, "]")?;
            }
        },
        expr::ExprKind::UseVar(e) => write!(f, "{}", e.name)?,
        expr::ExprKind::UseField(e) => {
            print_expr(f, i, &e.parent)?;
            write!(f, ".{}", e.name)?;
        }
        expr::ExprKind::UseIndex(e) => {
            print_expr(f, i, &e.parent)?;
            write!(f, "[")?;
            print_expr(f, i, &e.key)?;
            write!(f, "]")?;
        }
        expr::ExprKind::AssignVar(e) => {
            write!(f, "{}", e.name)?;
            match e.op {
                Some(op) => write!(f, " {op}= ")?,
                None => write!(f, " = ")?,
            }
            print_expr(f, i, &e.value)?;
        }
        expr::ExprKind::AssignField(e) => {
            print_expr(f, i, &e.parent)?;
            write!(f, ".{}", e.name)?;
            match e.op {
                Some(op) => write!(f, " {op}= ")?,
                None => write!(f, " = ")?,
            }
            print_expr(f, i, &e.value)?;
        }
        expr::ExprKind::AssignIndex(e) => {
            print_expr(f, i, &e.parent)?;
            write!(f, "[")?;
            print_expr(f, i, &e.key)?;
            write!(f, "]")?;
            match e.op {
                Some(op) => write!(f, " {op}= ")?,
                None => write!(f, " = ")?,
            }
            print_expr(f, i, &e.value)?;
        }
        expr::ExprKind::Call(e) => {
            print_expr(f, i, &e.callee)?;
            write!(f, "(")?;
            for arg in &e.args {
                if let Some(key) = &arg.key {
                    write!(f, "{key}: ")?;
                }
                print_expr(f, i, &arg.value)?;
                write!(f, ",")?;
            }
            write!(f, ")")?;
        }
        expr::ExprKind::MethodCall(e) => {
            print_expr(f, i, &e.receiver)?;
            write!(f, ".{}", e.method)?;
            write!(f, "(")?;
            for arg in &e.args {
                if let Some(key) = &arg.key {
                    write!(f, "{key}: ")?;
                }
                print_expr(f, i, &arg.value)?;
                write!(f, ",")?;
            }
            write!(f, ")")?;
        }
    }
    write!(f, "/*{}*/", e.ty)?;

    Ok(())
}
