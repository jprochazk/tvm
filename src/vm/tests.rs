use std::cell::RefCell;
use std::fmt::{Display, Write};

use super::*;
use crate::code::print::DisplayModule;
use crate::error::Error;
use crate::util::JoinIter as _;

fn report(e: Vec<Error>) -> String {
    e.into_iter().join("\n").to_string()
}

thread_local! {
    static EVENTS: RefCell<Vec<String>> = const { RefCell::new(Vec::new()) };
}

struct DisplayValue<'a>(&'a Value);
impl Display for DisplayValue<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self.0 {
            Value::Unit(_) => write!(f, "_"),
            Value::Bool(v) => write!(f, "{v}"),
            Value::I64(v) => write!(f, "{v}i"),
            Value::F64(v) => write!(f, "{v}i"),
        }
    }
}

impl Display for CallFrame {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "<frame {:?} base={}, ret={}>",
            self.callee.name, self.stack_base, self.ret_addr
        )
    }
}

impl Display for StackFrame<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0.iter().map(DisplayValue).join(", "))
    }
}

fn debug_hook(event: DebugEvent) {
    const ALIGN: usize = 24;

    EVENTS.with_borrow_mut(|events| match event {
        DebugEvent::Dispatch { op, stack_frame } => {
            let op = op.to_string();
            events.push(format!("{op:ALIGN$} [{stack_frame}]"));
        }
        DebugEvent::PostDispatch { stack_frame } => {
            let prev = events.last_mut().unwrap();
            write!(prev, " -> [{stack_frame}]").unwrap();
        }
        DebugEvent::Call { call_frame } => events.push(format!("{}", call_frame)),
    })
}

fn _run(input: &str) -> String {
    let ast = match crate::syn::try_parse(input) {
        Ok(ast) => ast,
        Err(e) => panic!("{}", report(e)),
    };
    let hir = match crate::ty::check(&ast) {
        Ok(hir) => hir,
        Err(e) => panic!("{}", report(e)),
    };
    let code = match crate::code::compile(hir) {
        Ok(code) => code,
        Err(e) => panic!("{}", report(e)),
    };

    let mut out = String::new();
    writeln!(
        &mut out,
        "## Module\n\n{}",
        DisplayModule(&code, Some(input))
    )
    .unwrap();

    write!(&mut out, "## Output\n\n").unwrap();
    match code.link().debug(debug_hook) {
        Ok(value) => {
            writeln!(&mut out, "{value:?}\n").unwrap();
        }
        Err(e) => {
            writeln!(&mut out, "{}\n", report(vec![e])).unwrap();
        }
    };

    EVENTS.with_borrow(|events| {
        writeln!(&mut out, "## Events\n\n{}", events.iter().join("\n")).unwrap();
    });

    out
}

macro_rules! run {
    ($input:literal) => {
        _run(indoc::indoc!($input))
    };
}

macro_rules! test {
    ($name:ident, $input:literal) => {
        #[test]
        fn $name() {
            insta::assert_snapshot!(run!($input))
        }
    };
}

test! {
    i64,
    r#"
        123
    "#
}

test! {
    f64,
    r#"
        1.23
    "#
}

test! {
    variable,
    r#"
        let v = 0;
    "#
}

test! {
    add_i64,
    r#"
        10 + 10
    "#
}

test! {
    arithmetic_i64,
    r#"
        10 + 10 * 20 - 10 / 2
    "#
}

test! {
    add_f64,
    r#"
        10.5 + 20.5
    "#
}

test! {
    arithmetic_f64,
    r#"
        1.23 + 3.21 * 6.12 - 3.0 / 1.5
    "#
}

test! {
    simple_call,
    r#"
        fn add(a: int, b: int) -> int { a + b }

        add(2, 3)
    "#
}

test! {
    nested_call,
    r#"
        fn id(v: int) -> int { v }
        fn add(a: int, b: int) -> int { id(a) + id(b) }

        add(id(2), id(3))
    "#
}
