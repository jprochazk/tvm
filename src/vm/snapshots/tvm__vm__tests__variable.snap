---
source: src/vm/tests.rs
expression: "emit! (r#\"\n        let v = 0;\n    \"#)"
---
Unit
